File: contests/abc237/E_Skiing.py
#              E - Skiing
# ----------------------------------------
# 問題
# https://atcoder.jp/contests/abc237/tasks/abc237_e
# 解説
# https://www.youtube.com/watch?v=HOIUaAO-aWc

# AC
# ----------------------------------------

# 負の辺にはダイクストラは使えない
# 負の辺に使えるベルマンフォード法はO(NM) => TLE
# なんとかしてダイクストラが使える形に落とし込みたい

# 1. 登るときにかかるコストだけを重みにしたグラフを作成(下りは0)
# 2. 最短経路を探索 (ダイクストラ法)
# 3. 求める答えは、H_goal - H_start + cost の最大値になる


from heapq import heappop, heappush
INF = 1e18

def dijkstra(G, s):
    N = len(G)
    dist = [INF] * N
    dist[s] = 0  # スタートの初期化

    q = [(0, s)]
    while q:
        w_old, u = heappop(q)

        if dist[u] < w_old:
            continue

        for v, w_new in G[u]:
            if dist[v] > dist[u] + w_new:
                dist[v] = dist[u] + w_new  # 緩和処理
                heappush(q, (dist[v], v))
    
    return dist

if __name__ == "__main__":
    N, M = map(int, input().split())
    H = list(map(int, input().split()))
    G = [[] for _ in range(N)]
    for i in range(M):
        u, v = map(int, input().split())
        u-=1; v-=1

        h, l = (u, v) if H[u] > H[v] else (v, u)

        # descend
        G[h].append((l, 0))   # cost: 0
        # climb
        G[l].append((h, H[h] - H[l]))  # cost: diff

    # 最短経路探索
    cost = dijkstra(G, 0)
    
    # (スタートとの標高差 - コスト)の最大値が答えとなる
    ans = [H[0]-height-cost for height, cost in zip(H, cost)]
    # print(ans)

    print(max(ans))

File: contests/abc237/E_Skiing.rs
//                E - Skiing               
// ----------------------------------------
// 問題
// https://atcoder.jp/contests/abc237/tasks/abc237_e
// ----------------------------------------

// attributes
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(non_snake_case)]
#![allow(dead_code)]
#![allow(unused_macros)]

// imports
use itertools::Itertools;
use proconio::{input, fastout, marker::{Chars, Bytes, Usize1}};
use std::collections::BinaryHeap;
use std::cmp::Reverse;

macro_rules! debug {
    ( $($val:expr),* $(,)* ) => {{
        #[cfg(debug_assertions)]
        eprintln!( concat!($(stringify!($val), " = {:?}, "),*), $($val),* );
    }};
}

fn debug_2d<T: std::fmt::Debug>(array: &Vec<Vec<T>>) {
    #[cfg(debug_assertions)]
    for row in array {
        println!("{:?}", row);
    }
}

// constant
const MOD1: usize = 1_000_000_007;
const MOD9: usize = 998_244_353;
const INF: usize = 1001001001001001001;

// Dijkstra法
fn dijkstra(graph: &Vec<Vec<(usize, isize)>>, s: usize) -> Vec<isize> {
    const INF: isize = 1001001001001001001;
    let n: usize = graph.len();
    let mut dist: Vec<isize> = vec![INF; n];
    let mut pq: BinaryHeap<Reverse<(isize, usize)>> = BinaryHeap::new();
    // 初期化
    dist[s] = 0;
    pq.push(Reverse((dist[s], s)));
    // 更新
    while let Some(Reverse((cost, u))) = pq.pop() {
        if dist[u] < cost { continue; }
        for &(v, weight) in &graph[u] {
            if dist[v] > dist[u] + weight {
                dist[v] = dist[u] + weight;
                pq.push(Reverse((dist[v], v)));
            }
        }
    }
    dist
}


/// ## 考察
/// - 辺の重みを低い方から高い方に，その絶対値分だけ付与する
/// - H0からHiに到達するとき，えられる楽しさの最大値は，
///   Hi - H0 - (A0からAiへの最短経路)
fn main() {
    input! {
        N: usize,
        M: usize,
        H: [isize; N],
        edges: [(Usize1, Usize1); M],
    }

    // グラフの構築
    let mut G = vec![vec![]; N];
    for &(mut u, mut v) in &edges {
        if H[u] < H[v] {
            std::mem::swap(&mut u, &mut v)
        }
        // 上→下（コストなし）
        G[u].push((v, 0));
        // 下→上（コストあり）
        G[v].push((u, H[u] - H[v]));
    }

    debug_2d(&G);

    // 最短経路探索
    let cost = dijkstra(&G, 0);

    // 答えを求める
    let mut ans = 0;
    for i in 0..N {
        ans = ans.max(
            H[0] - H[i] - cost[i]
        );
    }

    println!("{}", ans);
}
